hazelcast-sql/src/main/java/com/hazelcast/sql/impl/type/QueryDataType.java

```diff
@@ -148,6 +177,30 @@ public TypeKind getObjectTypeKind() {
         return objectTypeKind;
     }
 
+    public void addField(String name, QueryDataType type) {
+        if (objectFieldsBuilder == null) {
```

Comment:
I'm a little confused by this part. we don't have multithreading in this part?

Reply:
why not try creating a builder when creating a class, and then use the field objectFields  to understand the state

Reply:
> Why not try creating a builder when creating a class, and then use the field objectFields to understand the state?

The custom types created when parsing SQL DDL statements have no fields or metadata. During field validation, they are _enriched_, i.e. populated with fields and metadata. In that use case, the type is expected to be ready-to-use without calling `finalizeFields()`. The same can be achieved by adding a parameter to the constructor, or by calling `finalizeFields()` in the parser, but it would be counter-intuitive since the fields are not known at that stage.
> I'm a little confused by this part. we don't have multithreading in this part?

`QueryDataType` is thread-safe if no field will ever be added, or after the fields are finalized, as stated in the Javadoc.

Reply:
ok. Maybe think about adding call finalize() in the getObjectFields?

Reply:
If we do that, a list builder will be unnecessarily created for placeholder types.

Reply:
ok